source/common/upstream/transport_socket_match_impl.cc

```diff
@@ -75,5 +140,94 @@ TransportSocketMatcher::MatchData TransportSocketMatcherImpl::resolve(
   return {*default_match_.factory, default_match_.stats, default_match_.name};
 }
 
+TransportSocketMatcher::MatchData TransportSocketMatcherImpl::resolveUsingMatcher(
+    const envoy::config::core::v3::Metadata* endpoint_metadata,
+    const envoy::config::core::v3::Metadata* locality_metadata) const {
+  TransportSocketMatchingData data(endpoint_metadata, locality_metadata);
+  auto on_match = Matcher::evaluateMatch(*matcher_, data);
```

Comment:
So I'm confused here: I think transport socket matching happens after load balancing decision is made. How is downstream connection wired here? Where are the downstream addresses coming from? LBs in general do not have access to that information, it must be specifically requested via TransportSocketOptions (e.g. when forced with PROXY). If we don't intend to support downstream-connection based matching, maybe we should just add some generic filter-state/metadata one, so that filter state is explicitly passed from downstream to upstream, like we do with tunneling?

Reply:
Hmmm, yeah I think it's better to just use filter state here as you suggested. For our use-case where we need Network Namespace, I might need to raise another PR to get NNS as a well known network input.